src/mbgl/util/thread_pool.cpp

```diff
@@ -40,20 +49,83 @@ std::thread ThreadedSchedulerBase::makeSchedulerThread(size_t index) {
 
             auto function = std::move(queue.front());
             queue.pop();
+
+            if (function) {
+                pendingItems++;
+            }
+
             lock.unlock();
-            if (function) function();
+
+            if (function) {
+                const auto cleanup = [&] {
+                    // destroy the function and release its captures before unblocking `waitForEmpty`
+                    function = {};
+                    pendingItems--;
+                    if (queue.empty() && !pendingItems) {
+                        cvEmpty.notify_all();
+                    }
+                };
+                try {
+                    function();
+                    cleanup();
+                } catch (...) {
+                    lock.lock();
+                    if (handler) {
+                        handler(std::current_exception());
+                    }
+                    cleanup();
+                    if (handler) {
+                        continue;
+                    }
+                    throw;
+                }
+            }
         }
     });
 }
 
 void ThreadedSchedulerBase::schedule(std::function<void()>&& fn) {
     assert(fn);
-    {
-        std::lock_guard<std::mutex> lock(mutex);
-        queue.push(std::move(fn));
+    if (fn) {
+        {
+            // We need to block if adding adding a new task from a thread not controlled by this
+            // pool.  Tasks are added by other tasks, so we must not block a thread we do control
+            // or `waitForEmpty` will deadlock.
+            std::unique_lock<std::mutex> addLock(addMutex, std::defer_lock);
+            if (!isThreadOwned(std::this_thread::get_id())) {
+                addLock.lock();
+            }
+            std::lock_guard<std::mutex> lock(mutex);
+            queue.push(std::move(fn));
+        }
+        cvAvailable.notify_one();
     }
+}
 
-    cv.notify_one();
+std::size_t ThreadedSchedulerBase::waitForEmpty(Milliseconds timeout) {
+    // Must not be called from a thread in our pool, or we would deadlock
+    if (!isThreadOwned(std::this_thread::get_id())) {
```

Comment:
Thinking about this, why not store a pointer to the owner thread pool in TLS? Then we don't even need to check our own TID, just that the thread local pointer is a match. ie: (`util/thread_local.hpp`, `ThreadLocal<ThreadedSchedulerBase>`)

Reply:
Yeah, that's a good idea.